File: Tcf.rb
first = true
begin
    if first
        begin
            raise "Foo"
        rescue => detail
            puts "Inner Catch"
            raise detail
        ensure
            puts "Inner Finally"
        end
    end
rescue
    puts "Outer Catch"
    first = false
    puts "Retry fires the begin block again"
    retry
else
    puts "Else runs if there is no exceptions"
ensure
    puts "Outer Finally"
end

File: Tcf.py
from __future__ import print_function

try:
    try:
        try:
            raise Exception("Foo")
        except Exception as e:
            print('Inner Catch')
            raise
        finally:
            print('Inner Finally')
    except Exception as e:
        print('Outer Catch')
    finally:
        print('Outer Finally')
except:
    pass
else:
    print("Else runs if no exception is raised")
finally:
    print("Finally always runs")
